canonical/knative-rocks
# feat: add rock for `operator`

Closes https://github.com/canonical/bundle-kubeflow/issues/1132

Rock is based on the upstream ko image by `knative`: https://github.com/knative/operator/tree/knative-v1.12.4/cmd/operator

To compare the rock I used the upstream docker tag `knative-v1.12.4`, that is the version we are using in knative in kubeflow 1.9 [charm](https://github.com/canonical/knative-operators/blob/track/1.12/charms/knative-operator/metadata.yaml#L27).

## Testing with knative-operator charm
Prerequisite: setup juju+microk8s, see the [tutorial](https://charmed-kubeflow.io/docs/get-started) for instructions.
1. Download the Rock built from the PR Checks: https://github.com/canonical/knative-rocks/actions/runs/11798793786
2. Import the rock into your local docker registry using `skopeo`:
```
rockcraft.skopeo --insecure-policy copy oci-archive:knative-operator docker-daemon:knative-operator:test-rock
```
3. Import the rock into microk8s container registry
```
docker save knative-operator:test-rock-o knative-operator-rock.tar
microk8s ctr image import knative-operator-rock.tar
```
4. Replace the upstream image in the metadata.yaml with the local Rock. Diff:
```
-    upstream-source: gcr.io/knative-releases/knative.dev/operator/cmd/operator:v1.12.4
+    upstream-source: docker.io/library/knative-operator:test-rock
```
5. Run the knative bundle integration tests:
```
tox -vve integration -- --model knative-test --keep-models
```

Expected output:
```
======================================================= 7 passed in 599.54s (0:09:59) ========================================================
integration: 600494 I exit 0 (600.31 seconds) /home/nohaihab/repos/knative-operators> pytest --show-capture=no --log-cli-level=INFO -vvs --tb=native --model knative-test --keep-models tests/test_bundle.py pid=444477 [tox/execute/api.py:286]
  integration: OK (600.36=setup[0.05]+cmd[600.31] seconds)
  congratulations :) (600.41 seconds)
```
